File: GW3/LogFileHandling/Query/Tokens/TokenComa.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace GWLogger.Backend.DataContext.Query.Tokens
{
    class TokenComa : Token
    {
        public override bool CanBeUsed(QueryParser parser)
        {
            parser.SkipSpaces();
            return parser.PeekChar() == ',';
        }

        public override Token Extract(QueryParser parser)
        {
            parser.SkipSpaces();
            parser.NextChar();
            return new TokenComa();
        }
    }
}

File: GW3/GWLogger/Frontend/Query/Tokens/TokenComa.ts
class TokenComa extends Token {

    public tokenType: TokenType = TokenType.TokenComa;

    //public proposals: SuggestionInterface[] = <SuggestionInterface[]>[{ suggestion: "," }];
    public proposals: SuggestionInterface[] = []

    public canBeUsed(parser: QueryParser): boolean {
        parser.skipSpaces();
        return parser.peekChar() == ',';
    }

    public extract(parser: QueryParser): Token {
        parser.skipSpaces();
        parser.nextChar();
        let token = new TokenComa();
        token.value = ",";
        return token;
    }

    public getProposals(nextToken?: Token, afterNextToken?: Token): Token[] {
        if (typeof nextToken != 'undefined' && typeof afterNextToken != 'undefined') {
            switch (nextToken.tokenType) {
                case TokenType.TokenOperation:
                    return [new TokenOpenParenthesis()];
                case TokenType.TokenName:
                    return [
                        new TokenComa(),
                        new TokenGroup(),
                        new TokenLimit(),
                        new TokenOrder(),
                        new TokenWhere(),
                        new TokenString(),
                        new TokenName()
                    ];
                case TokenType.TokenStar:
                    return [
                        new TokenGroup(),
                        new TokenLimit(),
                        new TokenOrder(),
                        new TokenWhere()
                    ];
            }
        }
        return [];
    }
}